dot/telemetry/telemetry.go

```diff
@@ -57,126 +56,85 @@ var (
 		once.Do(
 			func() {
 				handlerInstance = &Handler{
-					buf: bytes.Buffer{},
+					msg: make(chan Message, 256),
+					log: log.New("pkg", "telemetry"),
 				}
-				log.SetOutput(&handlerInstance.buf)
-				log.SetFormatter(new(MyJSONFormatter))
-				go handlerInstance.sender()
+				go handlerInstance.startListening()
 			})
 	}
 	return handlerInstance
 }
 
-// AddConnections adds connections to telemetry sever
-func (h *Handler) AddConnections(conns []*genesis.TelemetryEndpoint) {
+// NewTelemetryMessage builds a telemetry message
+func NewTelemetryMessage(values ...*KeyValue) *Message { //nolint
+	mvals := make(map[string]interface{})
+	for _, v := range values {
+		mvals[v.key] = v.value
+	}
+	return &Message{
+		values: mvals,
+	}
+}
+
+// NewKeyValue builds a key value pair for telemetry messages
+func NewKeyValue(key string, value interface{}) *KeyValue { //nolint
+	return &KeyValue{
+		key:   key,
+		value: value,
+	}
+}
+
+// AddConnections adds the given telemetry endpoint as listeners that will receive telemetry data
+func (t *Handler) AddConnections(conns []*genesis.TelemetryEndpoint) {
 	for _, v := range conns {
 		c, _, err := websocket.DefaultDialer.Dial(v.Endpoint, nil)
 		if err != nil {
-			fmt.Printf("Error %v\n", err)
+			// todo (ed) try reconnecting if there is an error connecting
+			t.log.Warn("issue adding telemetry connection", err)
 			continue
 		}
-		h.wsConn = append(h.wsConn, c)
+		tConn := &telemetryConnection{
+			wsconn:    c,
+			verbosity: v.Verbosity,
+		}
+		t.connections = append(t.connections, tConn)
 	}
 }
 
-// ConnectionData struct to hold connection data
-type ConnectionData struct {
-	Authority     bool
-	Chain         string
-	GenesisHash   string
-	SystemName    string
-	NodeName      string
-	SystemVersion string
-	NetworkID     string
-	StartTime     string
-}
-
-// SendConnection sends connection request message to telemetry connection
-func (h *Handler) SendConnection(data *ConnectionData) {
-	h.Lock()
-	defer h.Unlock()
-	payload := log.Fields{"authority": data.Authority, "chain": data.Chain, "config": "", "genesis_hash": data.GenesisHash,
-		"implementation": data.SystemName, "msg": "system.connected", "name": data.NodeName, "network_id": data.NetworkID, "startup_time": data.StartTime,
-		"version": data.SystemVersion}
-	telemetryLogger := log.WithFields(log.Fields{"id": 1, "payload": payload, "ts": time.Now()})
-	telemetryLogger.Print()
-}
-
-// SendBlockImport sends block imported message to telemetry connection
-func (h *Handler) SendBlockImport(bestHash string, height *big.Int) {
-	h.Lock()
-	defer h.Unlock()
-	payload := log.Fields{"best": bestHash, "height": height.Int64(), "msg": "block.import", "origin": "NetworkInitialSync"}
-	telemetryLogger := log.WithFields(log.Fields{"id": 1, "payload": payload, "ts": time.Now()})
-	telemetryLogger.Print()
+// SendMessage sends Message to connected telemetry listeners
+func (t *Handler) SendMessage(msg *Message) {
+	t.msg <- *msg
```

Comment:
improvement: add a timeout and error return for this method.  I see that that it's a buffered channel of length 256, so this shouldn't happen, but it's a good practice so that this function call isn't blocking.

Reply:
Good suggestion, I've added timeout and return error.